plugins/modem-manager/fu-dfota-updater.c

```diff
@@ -0,0 +1,382 @@
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "firmware size mismatch - expected %" G_GSIZE_FORMAT
+			    ", but was %" G_GSIZE_FORMAT,
+			    size,
+			    size_parsed);
+		return FALSE;
+	}
+	if (g_strcmp0(checksum, checksum_parsed) != 0) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "checksum mismatch - expected %s, but was %s",
+			    checksum,
+			    checksum_parsed);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+gboolean
+fu_dfota_updater_open(FuDfotaUpdater *self, GError **error)
+{
+	if (self->io_channel == NULL) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "no channel provided for update");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+gboolean
+fu_dfota_updater_close(FuDfotaUpdater *self, GError **error)
+{
+	/* closing will be handled by locker */
+	return TRUE;
+}
+
+static gboolean
+fu_dfota_updater_parse_fota_response(gchar *response,
+				     FuProgress *progress,
+				     gboolean *finished,
+				     GError **error)
+{
+	g_autoptr(GRegex) fota_regex = NULL;
+	g_autoptr(GMatchInfo) match_info = NULL;
+	g_autofree gchar *status_match = NULL;
+	g_autofree gchar *status_number_match = NULL;
+	guint64 status_number = 0;
+
+	/* +QIND: "FOTA","<STATUS>"(,<number>)? */
+	fota_regex = g_regex_new("\\+QIND:\\s*\"FOTA\",\"([A-Z]+)\"(,(\\d+))?", 0, 0, NULL);
+
+	if (!g_regex_match(fota_regex, response, 0, &match_info)) {
+		g_warning("got unexpected response '%s'", response);
+		return TRUE;
+	}
+	if (!g_match_info_matches(match_info)) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "could not match fota status");
+		return FALSE;
+	}
+
+	status_match = g_match_info_fetch(match_info, 1);
+
+	if (g_strcmp0(status_match, "START") == 0) {
+		g_debug("update started successfully");
+		return TRUE;
+	}
+
+	/* expect status and number, which means four matches in the above regex */
+	if (g_match_info_get_match_count(match_info) != 4) {
+		g_warning("got badly formatted message '%s'", response);
+		return TRUE;
+	}
+
+	status_number_match = g_match_info_fetch(match_info, 3);
+	if (!g_ascii_string_to_unsigned(status_number_match,
+					10,
+					0,
+					G_MAXUINT,
+					&status_number,
+					error))
+		return FALSE;
+
+	if (g_strcmp0(status_match, "UPDATING") == 0) {
+		fu_progress_set_percentage(progress, (guint)status_number);
+		return TRUE;
+	}
+	if (g_strcmp0(status_match, "END") == 0) {
+		if (status_number != 0) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "update exited with error code %" G_GUINT64_FORMAT,
+				    status_number);
+			return FALSE;
+		}
+
+		g_debug("updated finished successfully");
+		*finished = TRUE;
+		return TRUE;
+	}
+
+	g_warning("got unhandled status '%s'", status_match);
+	return TRUE;
+}
+
+gboolean
+fu_dfota_updater_write(FuDfotaUpdater *self,
+		       FuProgress *progress,
```

Comment:
why are you passing `progress` into this function?

Reply:
`progress` is set here in `fu_dfota_updater_parse_fota_response`